lightning/src/onion_message/messenger.rs

```diff
@@ -434,17 +443,43 @@ pub enum Destination {
 }
 
 impl Destination {
+	/// Attempts to resolve the [`IntroductionNode::DirectedShortChannelId`] of a
+	/// [`Destination::BlindedPath`] to a [`IntroductionNode::NodeId`], if applicable, using the
+	/// provided [`ReadOnlyNetworkGraph`].
+	pub fn resolve(&mut self, network_graph: &ReadOnlyNetworkGraph) {
+		if let Destination::BlindedPath(path) = self {
+			let introduction_node = &mut path.introduction_node;
+			if let IntroductionNode::DirectedShortChannelId(direction, scid) = introduction_node {
+				let pubkey = network_graph
+					.channel(*scid)
+					.map(|c| match direction {
+						Direction::NodeOne => c.node_one,
+						Direction::NodeTwo => c.node_two,
+					})
+					.and_then(|node_id| node_id.as_pubkey().ok());
+				if let Some(pubkey) = pubkey {
+					*introduction_node = IntroductionNode::NodeId(pubkey);
+				}
+			}
+		}
```

Comment:
Since we have already defined `public_introduction_node_id` in BlindedPath. We can utilize it here to keep this function clean.

```suggestion
		if let Destination::BlindedPath(path) = self {
			if let Some(pubkey) = path
				.public_introduction_node_id(network_graph)
				.and_then(|node_id| node_id.as_pubkey().ok())
			{
				path.introduction_node = IntroductionNode::NodeId(pubkey);
			}
		}

```

On another note, `router.rs` defines `NodeId`, and `IntroductionNode` also has a variant `NodeId`. But these seem to be incompatible with each other, which doesn’t seem to me like ideal behavior.
Maybe we can think about `IntroductionNode::NodeId` taking `NodeId` as the field value, or maybe we can define an explicit conversion between them. Let me know what you think about this!

Reply:
> Since we have already defined `public_introduction_node_id` in BlindedPath. We can utilize it here to keep this function clean.

Good point, though we should only do this for the `DirectedShortChannelId`. If we do it for the `NodeId` variant then it could fail for an unannounced node even if the node is a peer.

> On another note, `router.rs` defines `NodeId`, and `IntroductionNode` also has a variant `NodeId`. But these seem to be incompatible with each other, which doesn’t seem to me like ideal behavior. Maybe we can think about `IntroductionNode::NodeId` taking `NodeId` as the field value, or maybe we can define an explicit conversion between them. Let me know what you think about this!

So generally, a node id is a pubkey. `NodeId` is simply a convenience type used to avoid `PublicKey`'s costly verification logic while deserializing a `NetworkGraph` (see #1107). That said, there should already be conversion functions on `NodeId` (e.g., the `as_pubkey` method used here).

Reply:
> If we do it for the NodeId variant then it could fail for an unannounced node even if the node is a peer.

Right! That makes sense!

> That said, there should already be conversion functions on NodeId (e.g., the as_pubkey method used here).

Yep, that's true!
But I was wondering if instead of `(gossip.rs) NodeId <-> pubkey <-> IntroductionNode::NodeId`, we can have a direct conversion between `(gossip.rs) NodeId <-> IntroductionNode::NodeId`.
But now I think about it, it would just be introducing extra complexity for not a lot of pay-off!